### DISCUSSION TITLE
Multi Partitioning - Challenge finding the best practice for my use case

### DISCUSSION THREAD

**Discussion by nirendy** [dagster-io/dagster#18465] [Q&A]
I'm new to this forum and have recently started exploring the Dagster framework. I'm genuinely impressed with its capabilities but have come across some challenges regarding MultiPartitions that I'd like to discuss and seek your insights on.

### The Challenge:
I am working with an external computed DataFrame asset, partitioned as follows: 
`(dynamic) product name -> daily file`.
While navigating through this, I've encountered certain features in MultiPartitions that seem to be either not implemented or perhaps I'm not utilizing established practices correctly.

### Expected goals from Dagster:
1 - Dagster recognizing the partitioning of this asset as an input to other assets
2 - Gain visibility of anomalies in the the data (can be missing file, or weird row/columns amount)
3 - Build on top of the existing data location, not needing to change/duplicate the data
4 - (Bonus) aggregation of summarize metadata across the products partitions

### Approaches I took:

1 - Creating a multi partition def with a dynamic + daily partitions (Seems to work fine)
I believe including this use case in the documentation would be beneficial. 
A nuance of my use case is that for each product, the daily partitioning start date is different, and I didn't find a way to achieve that.

2 - external asset
I haven't found support for partitioning in `external assets`. Although I found some functionality in the soon-to-be-deprecated `source asset`, it did not display the partition tab as effectively as in a standard `asset`.

3 - Creating an asset (with the partition from above), that opens the file (if exists) and save the shape in the metadata
a. The process was time-consuming across all partitions due to many subprocesses performing minor tasks. This is a minor issue since it only needs to run once, and I can probably optimize the queue configuration to improve it.
b. This approach also led to numerous small tasks for a single backfilling task (filling the runs log), which is again a minor concern as it's a one-time operation.
c. I experimented with both failing on missing files and reporting success with `exists=false` in the metadata

4 - Single run backfill policy (Go over the partitions, and manually create events of materializations)
 
a. A challenge was parsing the partition string myself, using "|" as a delimiter, which could be problematic if the string itself contains this character (not an issue in my case).

b. This partitioning caused the default `io_manager` to fail, so I created a dummy io-manager to override this behavior. However, this created an additional materialization event for each partition, superseding the event I had initially created.

c. Raising a `Failure` before exiting the asset
Every backfilling will be marked as failing which is also undesirable. 

5 - Instead of a backfill, create a job that creates materialization events.
I haven't tried this yet, but it may render the backfilling feature redundant, which is disappointing as I had hoped to utilize it effectively.

[Note: For points 2-5, the "plots" tab was not particularly useful in this context, though I had hoped it would assist in achieving Goal 2.]

6. I'm planning to try using `asset checks` to display plots that describes the anomalies in the data
My intention was to collect the metadata from each partition.
It seems to be harder than I thought, I didn't find any api that does it, the closest I found is `get_latest_materialization_events` but it gives information about the latest partition only, and `get_latest_storage_id_by_partition` but it doesn't return the event metadata. I implemented it by creating a combination of both, but I had to access private properties so I don't think that's a great solution.

I hope I was clear, and that my expectations from Dagster makes sense (hopefully I'm not trying to abuse Dagster for more that it is.  Maybe I should use another tool like mlflow to achieve some of that, but I hoped to achieve that in a single UI)

Thanks a lot for reading :)

**Comment by sryza** [comment]
> I haven't found support for partitioning in external assets. Although I found some functionality in the soon-to-be-deprecated source asset, it did not display the partition tab as effectively as in a standard asset.

Source assets are what I would recommend here. Would you be up for filing a github issue for any requests related to the partition tab and source assets? External assets are basically a renaming of this functionality and will get the same treatment.

> b. This partitioning caused the default io_manager to fail, so I created a dummy io-manager to override this behavior. However, this created an additional materialization event for each partition, superseding the event I had initially created.

Are you directly creating `AssetMaterialization` events? In general, we recommend against doing this, except in rare situations.

> It seems to be harder than I thought, I didn't find any api that does it, the closest I found is get_latest_materialization_events but it gives information about the latest partition only, and get_latest_storage_id_by_partition but it doesn't return the event metadata. I implemented it by creating a combination of both, but I had to access private properties so I don't think that's a great solution.

Take a look at https://github.com/dagster-io/dagster/discussions/14338.

> Every backfilling will be marked as failing which is also undesirable.

What would be desirable in this case?

**Reply by nirendy** [reply]
Thanks a lot for answering!

> Are you directly creating AssetMaterialization events? In general, we recommend against doing this, except in rare situations.

I'll clarify:
In this section of the docs [Launching single-run backfills using backfill policies](https://docs.dagster.io/concepts/partitions-schedules-sensors/backfills#launching-single-run-backfills-using-backfill-policies-) it suggested that `backfill_policy=BackfillPolicy.single_run()` is logically equivalent to running without it, but as far as I understand, it produce `AssetMaterialization` events differently. 
In my case for example, I want to provide a different metadata for each partition, and it's not possible to do so using that, hence the need to directly create `AssetMaterialization` event.

> Take a look at https://github.com/dagster-io/dagster/discussions/14338.

Thanks! that worked. I had to take a it one step further.
I wanted to get the metadata of all partitions latest materializations, so I used `get_latest_storage_id_by_partition` to do so:

```python
materialization = instance.get_event_records(
    event_records_filter=EventRecordsFilter(
        event_type=DagsterEventType.ASSET_MATERIALIZATION,
        storage_ids=list(instance.get_latest_storage_id_by_partition(
            event_type=DagsterEventType.ASSET_MATERIALIZATION,
            asset_key=AssetKey(["asset_1"])
        ).values())
    ),
)
```

> What would be desirable in this case?

Just to clarify, when I said "it's undesirable" I meant that the overall workaround solution I suggested is not desirable, and it's probably a result the `backfill_policy` workaround to store metadata in each partition.


> Source assets are what I would recommend here. Would you be up for filing a github issue for any requests related to the partition tab and source assets? External assets are basically a renaming of this functionality and will get the same treatment.

(Answering this lastly so I can refer to what I said above)

Correct me if I'm wrong:  You're suggesting to use the Source Asset, and in order to utilize the metadata, I still need to create materializations directly per partition, and insert the metadata to it.

I'll test that again and update soon.
I will be filling an issue once I'll test it and confirm that it's merely a UI issue.

**Reply by sryza** [reply]
> In my case for example, I want to provide a different metadata for each partition, and it's not possible to do so using that, hence the need to directly create AssetMaterialization event.

Here's an issue that tracks this limitation: https://github.com/dagster-io/dagster/issues/18123
